File: 백준/Gold/1941. 소문난 칠공주/소문난 칠공주.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;

public class Main {

    static final int N = 5;

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int[] pick = new int[7];
    static char[][] arr;

    public static void main(String[] args) throws IOException {

        arr = new char[N][N];

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int i = 0; i < N; i++) {
            arr[i] = br.readLine().toCharArray();
        }

        System.out.println(getAns(0, 0, 0, 0));
    }

    private static int getAns(int start, int depth, int countY, int countS) {

        if (countY > 3) return 0;

        if (depth == 7) {
            return (countS >= 4 && bfs()) ? 1 : 0;
        }

        int ans = 0;

        for (int i = start; i < N * N; i++) {

            pick[depth] = i;

            int x = i / N;
            int y = i % N;

            if (arr[x][y] == 'Y') {
                ans += getAns(i + 1, depth + 1, countY + 1, countS);
            } else {
                ans += getAns(i + 1, depth + 1, countY, countS + 1);
            }
        }

        return ans;
    }

    private static boolean bfs() {

        Queue<Integer> qu = new ArrayDeque<>();
        boolean[] selected = new boolean[N * N];
        boolean[] visit = new boolean[N * N];

        for (int p : pick) {
            selected[p] = true;
        }

        int start = pick[0];
        visit[start] = true;
        qu.offer(start);

        int connectedCount = 1;

        while (!qu.isEmpty()) {

            int now = qu.poll();
            int x = now / N;
            int y = now % N;

            for (int i = 0; i < 4; i++) {

                int nx = x + dx[i];
                int ny = y + dy[i];
                int next = nx * N + ny;

                if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;

                if (selected[next] && !visit[next]) {
                    visit[next] = true;
                    qu.offer(next);
                    connectedCount++;
                }
            }
        }

        return connectedCount == 7;
    }
}

File: 백준/Gold/1941. 소문난 칠공주/소문난 칠공주.py
from collections import deque
from itertools import combinations

N = 5
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

arr = [list(input().strip()) for _ in range(N)]


def bfs(pick):

    visit = set()
    qu = deque([pick[0]])
    visit.add(pick[0])

    connected_count = 1

    while qu:
        x, y = divmod(qu.popleft(), N)

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            next = nx * N + ny

            if nx < 0 or ny < 0 or nx >= N or ny >= N:
                continue

            if next in pick and next not in visit:
                visit.add(next)
                qu.append(next)
                connected_count += 1

    return connected_count == 7


ans = 0

for comb in combinations(range(25), 7):
    count_s = 0
    pick = []

    for idx in comb:
        x, y = divmod(idx, N)
        pick.append(idx)
        if arr[x][y] == 'S':
            count_s += 1

    if count_s >= 4 and bfs(pick):
        ans += 1

print(ans)